# GGS416 Satellite Image Analysis - Week 6

In this tutorial we are going to explore the basic functionality of OpenCV: https://opencv.org/

OpenCV stands for Open Source Computer Vision Library: https://github.com/opencv/opencv





In [ ]:
# Install OpenCV if you don't already have it.
import sys
!{sys.executable} -m pip install opencv-python

In [32]:
import cv2

# Load an image using 'imread' specifying the path to image
my_image = cv2.imread('../week4/clipped.tif')
my_image

# To display our image variable, we use 'imshow'
# The first parameter will be title shown on image window
# The second parameter is the image varialbe
cv2.imshow('Hello World', my_image)

# 'waitKey' allows us to input information when a image window is open
# By leaving it blank it just waits for anykey to be pressed before 
# continuing. By placing numbers (except 0), we can specify a delay for
# how long you keep the window open (time is in milliseconds here)
cv2.waitKey()

# This closes all open windows 
# Failure to place this will cause your program to hang
cv2.destroyAllWindows()

### Let's take a closer look at how images are stored




In [7]:
# Import numpy
import numpy as np
print(my_image.shape)

(4684, 9385, 3)


#### Shape gives the dimensions of the image array

The 2D dimensions are 4684 pixels in height bv 9385 pixels wide.
The '3L' means that there are 3 other components (RGB) that make up this image.

## Saving images using OpenCV

In [9]:
# Simply use 'imwrite' specificing the file name and the image to be saved
cv2.imwrite('output.jpg', my_image)
cv2.imwrite('output.png', my_image)

True

## Grayscaling

#### Grayscaling is process by which an image is converted from a full color to shades of grey (black & white)

In OpenCV, many functions grayscale images before processing. This is done because it simplifies the image, acting almost as a noise reduction and increasing processing time as there is less information in the image.

### Let convert our color image to greyscale

In [14]:
# Load our input image
# image = cv2.imread('./images/abe.jpg')
cv2.imshow('Original', my_image)
cv2.waitKey()

# We use cvtColor, to convert to grayscale
gray_image = cv2.cvtColor(my_image, cv2.COLOR_BGR2GRAY)

cv2.imshow('Grayscale', gray_image)
cv2.waitKey()
cv2.destroyAllWindows()

## Let's take a closer look at color spaces

You may have remembered we talked about images being stored in RGB (Red Green Blue) color Spaces. Let's take a look at that in OpenCV.

### First thing to remember about OpenCV's RGB is that it's BGR (I know, this is annoying)

### Let's look at the individual color levels for the first pixel (0,0)

In [24]:
# BGR Values for the first 0,0 pixel
blue, green, red = my_image[0, 0] 
print(blue, green, red)
print(my_image[0, 0])

print(my_image.shape)

0 0 0
[0 0 0]
(4684, 9385, 3)


Let's see what happens when we convert it to grayscale

In [22]:
gray_img = cv2.cvtColor(my_image, cv2.COLOR_BGR2GRAY)
print(gray_img.shape)

(4684, 9385)


It's now only 2 dimensions. Each pixel coordinate has only one value (previously 3) with a range of 0 to 255

In [23]:
gray_img[0, 0] 

0

### Another useful color space is HSV 
Infact HSV is very useful in color filtering.

In [30]:
#Hue: 0 - 180, Saturation: 0 - 255, Value: 0 - 255

image = cv2.imread('../week4/clipped.tif')

hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)

cv2.imshow('HSV image', hsv_image)
cv2.imshow('Hue channel', hsv_image[:, :, 0])
cv2.imshow('Saturation channel', hsv_image[:, :, 1])
cv2.imshow('Value channel', hsv_image[:, :, 2])

cv2.waitKey()
cv2.destroyAllWindows()

### Let's now explore lookng at individual channels in an RGB image

In [29]:
image = cv2.imread('../week4/clipped.tif')

# OpenCV's 'split' function splites the image into each color index
B, G, R = cv2.split(image)

print(B.shape)
cv2.imshow("Red", R)
cv2.imshow("Green", G)
cv2.imshow("Blue", B)
cv2.waitKey(0)
cv2.destroyAllWindows()

# Let's re-make the original image, 
merged = cv2.merge([B, G,R]) 
cv2.imshow("Merged", merged) 

# Let's amplify the blue color
merged = cv2.merge([B+100, G, R])
cv2.imshow("Merged with Blue Amplified", merged) 

cv2.waitKey(0)
cv2.destroyAllWindows()

(1562, 3129)


In [34]:
import cv2
import numpy as np

B, G, R = cv2.split(my_image)

# Let's create a matrix of zeros 
# with dimensions of the image h x w  
zeros = np.zeros(image.shape[:2], dtype = "uint8")

cv2.imshow("Red", cv2.merge([zeros, zeros, R]))
cv2.imshow("Green", cv2.merge([zeros, G, zeros]))
cv2.imshow("Blue", cv2.merge([B, zeros, zeros]))

cv2.waitKey(0)
cv2.destroyAllWindows()